### PR TITLE
Implement File Stats and Segment elimination

### PR DESCRIPTION
**Pull request by zhousun**
Store Column Statistics for each file in pg catalog table, and use them in scan to skip files.

Note: the feature covers most use cases of partitioned tables, after the change pg_mooncake would work very well for timeseries use cases.

### CODE REVIEW

File: src/columnstore/columnstore_stats.hpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include "duckdb/catalog/catalog_entry/table_catalog_entry.hpp"
+
+namespace duckdb {
+struct ColumnstoreStats {
```
Comment (dpxcc) [inline src/columnstore/columnstore_stats.hpp:6]: Considering using DuckDB's ObjectCache

File: src/columnstore/columnstore_stats.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    static ColumnstoreStats Deserialize(Deserializer &deserializer);
+
+    map<string, unique_ptr<BaseStatistics>> stats_map;
+};
+
+struct ColumnstoreStatsMap {
+    void LoadStats(const string file, const char *data, int len);
+
+    BaseStatistics *GetStats(const string path, const string col) {
```
Comment (dpxcc) [inline src/columnstore/columnstore_stats.hpp:21]: string -> string &

File: src/columnstore/columnstore_table.hpp [hunk 3]
```diff
@@ -9,6 +9,18 @@ class ColumnstoreMetadata;
 class ColumnstoreWriter;
 class DataChunk;
 
+struct ColumnstoreStats {
+    void AddStats(const string &column, BaseStatistics &stats);
```
Comment (dpxcc) [inline src/columnstore/columnstore_table.hpp]: need to also remove stats when data files are deleted

**Comment by dpxcc** [comment]
This is awesome!

**Comment by dpxcc** [comment]
I've cleaned up the implementation and addressed all the issues above. The main changes are
1. We serialize Parquet's metadata instead of a custom stats object
2. Support non-numeric stats

**Comment by dpxcc** [comment]
DuckDB is already doing row group level elimination in `ParquetReader::PrepareRowGroupBuffer()` https://github.com/duckdb/duckdb/blob/main/extension/parquet/parquet_reader.cpp#L764

What we did is to add a fast path:
1. We store file metadata in Postgres table so that DuckDB doesn't need to download it
2. We parse file metadata and aggregate file level stats to also save some time

**Comment by YuweiXiao** [comment]
> DuckDB is already doing row group level elimination in `ParquetReader::PrepareRowGroupBuffer()` https://github.com/duckdb/duckdb/blob/main/extension/parquet/parquet_reader.cpp#L764
> 
> What we did is to add a fast path:
> 
> 1. We store file metadata in Postgres table so that DuckDB doesn't need to download it
> 2. We parse file metadata and aggregate file level stats to also save some time

Thanks @dpxcc !

**Pull request closed without merging**